src/Pos/Wallet/Web/Server/Methods.hs

```diff
@@ -711,6 +732,60 @@ prepareTxInfo passphrase moneySource dstDistr = do
     listF separator formatter =
         F.later $ fold . intersperse separator . fmap (F.bprint formatter)
 
+    unknownFeePolicy =
+        InternalError . sformat ("Unknown Fee Policy, tag: "%build)
+
+-- Accept all addresses in decrease order (by coins)
+-- Destination addresses
+-- Sum of destination addresses
+-- Approximate fee for buildable transaction
+selectSrcAddresses
+    :: WalletWebMode m
+    => [(CWAddressMeta, Coin)]
+    -> Coin
+    -> TxFee
+    -> m (Coin, NonEmpty (CWAddressMeta, Coin))
+selectSrcAddresses allAddrs outputCoins (TxFee fee) =
+    either (throwM . RequestError) pure $
+    selectSrcAddressesDo (outputCoins `unsafeAddCoin` fee) allAddrs
+  where
+    selectSrcAddressesDo
+        :: WalletWebMode m => Coin
+        -> [(CWAddressMeta, Coin)]
+        -> Either Text (Coin, NonEmpty (CWAddressMeta, Coin))
+    selectSrcAddressesDo reqCoins addresses
+        | reqCoins == mkCoin 0 =
+            Left "Spending non-positive amount of money!"
+        | [] <- addresses =
+            Left $ sformat ("Not enough money (need " %build % " more)") reqCoins
+        | (ad, balance):addrs <- addresses = do
+            if | balance == mkCoin 0 ->
+                   selectSrcAddressesDo reqCoins addrs
+               | balance < reqCoins -> do
+                   let remCoins = reqCoins `unsafeSubCoin` balance
+                   ((ad, balance) :|) . toList <<$>> selectSrcAddressesDo remCoins addrs
+               | otherwise ->
+                   -- When balance >= reqCoins,
+                   -- then lets try to find input with exactly @reqCoins@ coins,
+                   -- in order to use one address instead of two.
+                   maybe (Right (balance `unsafeSubCoin` reqCoins, (ad, reqCoins) :| []))
+                         (\fa -> Right (mkCoin 0, fa :| []))
+                         (find ((reqCoins ==) . snd) addresses)
+
+withSafeSigners
+    :: (MonadIO m, MonadThrow m)
+    => PassPhrase
+    -> NonEmpty (EncryptedSecretKey)
+    -> (NonEmpty SafeSigner -> m a) -> m a
+withSafeSigners passphrase (sk :| sks) action =
+    withSafeSigner sk (pure passphrase) $ \mss -> do
+        ss <- maybeThrow (RequestError "Passphrase doesn't match") mss
+        case nonEmpty sks of
+            Nothing -> action (ss :| [])
+            Just sks' -> do
+                let action' = action . (ss :|) . toList
+                withSafeSigners passphrase sks' action'
```

Comment:
the main reason not to do it: this function depends on `WalletError` which is in `Pos.Wallet.Web.Error.Types` module, so we would get cyclic deps